### PR TITLE
More flexible partisans

### PR DESCRIPTION
**Pull request by lmoureaux**
This PR adds flexibility to partisans and changes the default behavior when they cost population:

* Multiple partisan unit types are now supported. The game will pick one the player can build, or the first one if the player can't build any
* When partisan units cost population, the population is taken away from the conquered city (rationale: it's an exodus)

## Testing

In the ruleset you're playing:

* Lower or remove requirements for the `Inspire_Partisans` effect
* Add the `Partisan` flag to some normal unit, e.g. Musketeers
* Change your favorite Partisan unit's `pop_cost`

Then conquer a city. You should see partisans of an appropriate unit type appear and the city size be reduced accordingly if they cost population. The city should never be wiped out.

### CODE REVIEW

File: server/unittools.cpp [hunk 1]
```diff
@@ -1176,7 +1176,15 @@ void place_partisans(struct tile *pcenter, struct player *powner, int count,
                      int sq_radius)
 {
   struct tile *ptile = nullptr;
-  struct unit_type *u_type = get_role_unit(L_PARTISAN, 0);
+  struct unit_type *u_type = best_role_unit_for_player(powner, L_PARTISAN);
+  if (!u_type) {
+    // If the player cannot build any partisan, pick the first one
```
Comment (blabber) [inline server/unittools.cpp:1181]: I guess this should read `// If the player can build any partisan, pick the first one`?

**Comment by blabber** [comment]
Deleted my comment after rereading the code. Sorry for the noise.

File: docs/Modding/Rulesets/effects.rst [hunk 2]
```diff
@@ -809,6 +809,14 @@ Has_Senate
 
 Inspire_Partisans
     Partisan units (defined in :file:`units.ruleset`) may spring up when this player's cities are taken.
+    The number of partisans created is a function of the city size.
+    When multiple unit types have the Partisan flag, the game tries to choose one the player can build. If
+    the player cannot build any, the first Partisan unit type is used. If the created partisans units have a
+    non-zero ``pop_cost``, the city will lose citizens accordingly.
+
+    Note that this effect can create units that the player could normally not build (because the tech hasn't
```
Comment (jwrober) [inline docs/Modding/Rulesets/effects.rst:817]: Make this an RST note `.. note::`?

**Review by jwrober — CHANGES_REQUESTED** [review]
One small edit to docs. I do want to test before a full approval. Looks great so far.

**Review by jwrober — APPROVED** [review]
Other than the small docs change, rest looks great. Was able to follow steps in a test game.

File: docs/Modding/Rulesets/effects.rst (hunk 2 above)
Comment (lmoureaux) [reply docs/Modding/Rulesets/effects.rst]: Updated:
image

**Review by jwrober — APPROVED** [review]

**Pull request merged**